### PR TITLE
feat: hooks round 3 - begin adding the Before and After user created hooks

### PR DESCRIPTION
**Pull request by cstockton**
## Hooks Round 3

This PR will contain a series of commits preparing for the implementation of before & after user created hooks. It takes the feedback from https://github.com/supabase/auth/pull/2012 into consideration.

### Summary

Begin adding the Before and After user created hooks:
* update pkg `internal/conf` [d5f5436](https://github.com/supabase/auth/pull/2028/commits/d5f5436ce173d3973fd32c9f970e57e739ae90f6)
    * add `BeforeUserCreated` and `AfterUserCreated` to `HookConfiguration` struct
    * add test cases for `EmailValidationBlockedMX` to restore 100% test coverage
* update pkg `internal/hooks/v0hooks` [bd37fe2](https://github.com/supabase/auth/pull/2028/commits/bd37fe23cb784939a81f782b533e4fe0247283f5)
    * add `BeforeUserCreated` method to `v0hooks.Manager` struct
    * add `AfterUserCreated` method to `v0hooks.Manager` struct
    * add tests to reach 100% coverage
* add pkg `internal/e2e/e2ehooks` [903e623](https://github.com/supabase/auth/pull/2028/commits/903e623ea110add81c40bb6ff6832f43dd53fb2f)
    * add `HookCall` to record calls to hooks
    * add `Hook` struct to hold `[]*HookCall` for a given hook name
    * add `HookRecorder` to hold one `Hook` object per hook name
    * add `Instance` struct to hold the `httptest.Server` and `HookRecorder`
    * add `AfterUserCreated` method to `v0hooks.Manager` struct
    * add tests to reach 100% coverage in all `internal/e2e` packages
* update pkg `internal/hooks/v0hooks` [829aec6](https://github.com/supabase/auth/pull/2028/commits/829aec6bb9371bfb9923a848ecac4dd7b3235bf3)
    * fix struct and json tag to match to match the Metadata type
* update pkg `internal/hooks/v0hooks` [ca67be0](https://github.com/supabase/auth/pull/2028/commits/ca67be0db26bd096697de8b0b53346791f3a3268)
    * remove `BeforeUserCreated` and `AfterUserCreated` methods
    * add Before & After user created hooks in `InvokeHook`
* update pkg `internal/e2e/e2eapi` [46c144e](https://github.com/supabase/auth/pull/2028/commits/46c144e1ea9eac8019621d6942d2166f74c5c7fd)
    * add comments in IOError tests involving `http.RoundTripper`
    * update calls to `t.Fatal` to use `require`

### Depends on
[feat: hooks round 1](https://github.com/supabase/auth/pull/2023) - prepare package structure
* Renamed pkg `internal/hooks/v0hooks/v0http` -> `internal/hooks/hookshttp` [8a398ab](https://github.com/supabase/auth/pull/2023/commits/8a398aba564267e4caa4c32bf661699d47d29174)
* Renamed pkg `internal/hooks/v0hooks/v0pgfunc` -> `internal/hooks/hookspgfunc` [8a398ab](https://github.com/supabase/auth/pull/2023/commits/8a398aba564267e4caa4c32bf661699d47d29174)
* Use pkg `internal/e2e` for test setup in:
    * pkg `internal/hooks/hookspgfunc` [4d60288](https://github.com/supabase/auth/pull/2023/commits/4d6028869f6303321571f0978a4151f4747f9c31)
    * pkg `internal/hooks/v0hooks` [4a7432b](https://github.com/supabase/auth/pull/2023/commits/4a7432b66a767c57d25423b1e13708c47cc3a69a)

[feat: hooks round 2](https://github.com/supabase/auth/pull/2025) - remove indirection and simplify error handling
* update pkg `internal/api` to:
    * uses `internal/hooks/v0hooks.Manager` instead of `internal/hooks/hooks.Manager` [aec5995](https://github.com/supabase/auth/pull/2025/commits/aec59956b1c68ac4bbcaa656251a3085bc64e551)
* remove pkg `internal/hooks/hooks.Manager` [062da5d](https://github.com/supabase/auth/pull/2025/commits/062da5da58c95552c2312d1d2709486226613c8f)
* add pkg `internal/hooks/hookserrors` [7e80afc](https://github.com/supabase/auth/pull/2025/commits/7e80afc355322f3970a7c7715cdbde7a144c716d)
* use pkg `internal/hooks/hookserrors` in `internal/hooks/v0hooks` [57744e8](https://github.com/supabase/auth/pull/2025/commits/57744e8c4136d54d77a7520242a98ff5c3b7799d)
* update pkg `internal/hooks/v0hooks` with an `Enabled` method [16cc4c9](https://github.com/supabase/auth/pull/2025/commits/16cc4c912475f7df632f958628fca49cfba482e1)

### CODE REVIEW

File: internal/hooks/v0hooks/v0hooks.go [hunk 1]
```diff
@@ -15,13 +19,76 @@ const (
...
+	}
+}
+
+type BeforeUserCreatedOutput struct {
+	Decision string `json:"decision"`
+	Message  string `json:"message"`
+}
+
+type AfterUserCreatedInput struct {
+	Header *Metadata    `json:"header"`
```
Comment (hf) [inline internal/hooks/v0hooks/v0hooks.go:76]:
```suggestion
	Metadata *Metadata    `json:"metadata"`
```

File: internal/hooks/v0hooks/v0hooks.go [hunk 2]
```diff
@@ -15,13 +19,76 @@ const (
...
+	return &Metadata{
+		UUID:      uuid.Must(uuid.NewV4()),
+		Time:      time.Now(),
+		IPAddress: utilities.GetIPAddress(r),
+		Name:      name,
+	}
+}
+
+type BeforeUserCreatedInput struct {
+	Header *Metadata    `json:"header"`
```
Comment (hf) [inline internal/hooks/v0hooks/v0hooks.go:56]:
```suggestion
	Metadata *Metadata    `json:"metadata"`
```

File: internal/hooks/v0hooks/manager.go [hunk 3]
```diff
@@ -58,11 +58,33 @@ func configByName(
...
+
+func (o *Manager) AfterUserCreated(
+	ctx context.Context,
+	tx *storage.Connection,
+	req *AfterUserCreatedInput,
+	res *AfterUserCreatedOutput,
+) error {
+	return o.dispatch(ctx, &o.config.Hook.AfterUserCreated, tx, req, res)
+}
+
```
Comment (hf) [inline internal/hooks/v0hooks/manager.go:87]: Why aren't these inside `InvokeHook`?

The tests can be changed to use `InvokeHook` instead with the correct inputs & outputs. `InvokeHook` can also make sure the `Metadata` struct field is filled in with the correct information.

**Comment by hf** [comment]
(jfyi If you merge now it will merge to the other PR which is already approved, so you can yolo this but please don't)

File: internal/hooks/v0hooks/manager.go (hunk 3 above)
Comment (cstockton) [reply internal/hooks/v0hooks/manager.go:87]: I would like to enforce strong typing when I begin calling these methods from the API package. I would like to remove the InvokeHook method in a future commit to remove the need for the type assertions.

Thoughts?

File: internal/hooks/v0hooks/v0hooks.go (hunk 2 above)
Comment (cstockton) [reply internal/hooks/v0hooks/v0hooks.go:56]: Good catch thank you

File: internal/hooks/v0hooks/v0hooks.go (hunk 1 above)
Comment (cstockton) [reply internal/hooks/v0hooks/v0hooks.go:76]: Good catch thank you

File: internal/hooks/v0hooks/manager.go (hunk 3 above)
Comment (hf) [reply internal/hooks/v0hooks/manager.go:87]: Let's do that for all of the hooks in a future PR, but maybe start a Slack discussion on it first. How is it going to look like exactly? 

Right now the `input`/`output` are `any` but it's not like you can accidentally pass an unsupported value -- a programmer error either way. Maybe generics can be used?

Let's discuss on Slack

Comment (hf) [reply internal/hooks/v0hooks/manager.go:87]: Polymorphism can also be used by refactoring Input & Output into, say, `CustomAccessTokenHookInvocation` struct that implements an `Invoke()` interface that uses `ctx, tx, ...`

Comment (cstockton) [reply internal/hooks/v0hooks/manager.go]: I've updated PR to have them inside InvokeHook.

File: internal/hooks/v0hooks/v0hooks.go [hunk 4]
```diff
@@ -15,13 +19,76 @@ const (
...
+
+func NewAfterUserCreatedInput(
+	r *http.Request,
+	user *models.User,
+) *AfterUserCreatedInput {
+	return &AfterUserCreatedInput{
+		Metadata: NewMetadata(r, AfterUserCreated),
+		User:     user,
+	}
+}
```
Comment (hf) [inline internal/hooks/v0hooks/v0hooks.go:88]: Can't we just use the zero struct values, cutting down on more code paths to test?

File: internal/e2e/e2ehooks/e2ehooks_test.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+			}
+
+			var got int
+			if err := call.Unmarshal(&got); err != nil {
+				t.Fatalf("exp nil err; got %v", err)
+			}
+			if exp := 12345; exp != got {
+				t.Fatalf("exp %v; got %v", exp, got)
+			}
+		}
```
Comment (hf) [inline internal/e2e/e2ehooks/e2ehooks_test.go:171]: Wrap this in `Run(func (t *testing.T) {})` to be clear which one fails.

Wouldn't hurt to use the `require` / `assert` packages in general to test expectations -- as used in all the other tests in the codebase.

File: internal/e2e/e2ehooks/e2ehooks_test.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+			t.Fatalf("exp nil err; got %v", err)
+		}
+		if inst == nil {
+			t.Fatal("exp non-nil *Instance")
+		}
+		if err := inst.Close(); err != nil {
+			t.Fatalf("exp nil err from Close; got %v", err)
+		}
+	}
+}
```
Comment (hf) [inline internal/e2e/e2ehooks/e2ehooks_test.go:51]: This test also doesn't add that much value. If `New` failed I'm pretty sure we wouldn't be able to release anything 😅 

The expectations also test non-nil, no error... pretty expected.

File: internal/hooks/v0hooks/v0hooks.go (hunk 4 above)
Comment (cstockton) [reply internal/hooks/v0hooks/v0hooks.go:88]: Well the zero values are not valid, the main purpose of the New is to ensure the [Metadata](https://github.com/supabase/auth/pull/2028/files/ca67be0db26bd096697de8b0b53346791f3a3268#diff-070d7377dd35255e1da96c9252a425b4f731c54d85f3c7675bd3a3cff522e5c3R46) is setup. Without them every single place we invoke a hook will look like:
```Go
	input := &BeforeUserCreatedInput{
		Metadata: &Metadata{
			UUID:      uuid.Must(uuid.NewV4()),
			Time:      time.Now(),
			IPAddress: utilities.GetIPAddress(r),
			Name:      BeforeUserCreated,
		},
		User: user,
	}
```

File: internal/e2e/e2eapi/e2eapi_test.go [hunk 7]
```diff
@@ -116,4 +120,56 @@ func TestDo(t *testing.T) {
...
+			})
+
+			ts := httptest.NewServer(hr)
+			defer ts.Close()
+
+			err := Do(ctx, http.MethodPost, ts.URL, nil, nil)
+			require.Error(t, err)
+			require.Equal(t, sentinel, err)
+		}()
+	}
```
Comment (hf) [inline internal/e2e/e2eapi/e2eapi_test.go:168]: I'm reading this for 5 minutes and I can't see what is it doing? Can you clean it up so it's either more obvious, or add some comments.

File: internal/e2e/e2ehooks/e2ehooks_test.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+		u := "http://localhost/hooks/__invalid-hook-name__"
+		rdr := strings.NewReader("12345")
+		req := httptest.NewRequestWithContext(ctx, "POST", u, rdr)
+		res := httptest.NewRecorder()
+		hookRec.ServeHTTP(res, req)
+
+		if exp, got := 404, res.Result().StatusCode; exp != got {
+			t.Fatalf("exp %v; got %v", exp, got)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (cstockton) [reply internal/e2e/e2ehooks/e2ehooks_test.go:185]: I disagree, It takes coverage from `98.4%` to `100%`.

File: internal/hooks/v0hooks/v0hooks.go (hunk 4 above)
Comment (hf) [reply internal/hooks/v0hooks/v0hooks.go:88]: Or...

```go
	input := &BeforeUserCreatedInput{
		Metadata: NewMetadataFromRequest(r, BeforeUserCreated),
		User: user,
	}
```

File: internal/e2e/e2ehooks/e2ehooks_test.go (hunk 6 above)
Comment (cstockton) [reply internal/e2e/e2ehooks/e2ehooks_test.go:51]: The value is that It takes test coverage from `96.8%` to `100%`

File: internal/hooks/v0hooks/v0hooks.go (hunk 4 above)
Comment (hf) [reply internal/hooks/v0hooks/v0hooks.go:88]: Actually in general I prefer if the `Metadata` is set as part of `InvokeHook` so callers only fill in the relevant payload.

File: internal/hooks/v0hooks/manager.go [hunk 9]
```diff
@@ -125,6 +129,23 @@ func (o *Manager) invokeHook(
...
+		}
+		return o.dispatch(
+			r.Context(), &o.config.Hook.BeforeUserCreated, conn, input, output)
+
+	case *AfterUserCreatedInput:
+		_, ok := output.(*AfterUserCreatedOutput)
+		if !ok {
+			return apierrors.NewInternalServerError(
+				"output should be *hooks.AfterUserCreatedOutput")
+		}
```
Comment (hf) [inline internal/hooks/v0hooks/manager.go:146]: These errors should be `panic` honestly, it's our programmer error. I don't know why they were created like this in the past... (no preference, just a note)

File: internal/e2e/e2ehooks/e2ehooks_test.go (hunk 6 above)
Comment (hf) [reply internal/e2e/e2ehooks/e2ehooks_test.go:51]: Yea I don't think it's worth adding 30 LOC to cover 10 in this case

File: internal/hooks/v0hooks/manager.go (hunk 9 above)
Comment (cstockton) [reply internal/hooks/v0hooks/manager.go:146]: My preference here is still to remove the need for these checks all together by having the Input and Output as part of the method signatures instead of everything using InvokeHook. Does seeing this code make you more open to that idea in a future pull request?

File: internal/hooks/v0hooks/v0hooks.go (hunk 4 above)
Comment (cstockton) [reply internal/hooks/v0hooks/v0hooks.go:88]: New* funcs to create objects is an established pattern in Go and we use it all over the code base, are we sure we want to [debate these minor points](https://en.wikipedia.org/wiki/Law_of_triviality)?

File: internal/hooks/v0hooks/manager.go (hunk 9 above)
Comment (hf) [reply internal/hooks/v0hooks/manager.go:146]: Yea ofc it does, was just commenting on it for future consideration

File: internal/e2e/e2ehooks/e2ehooks_test.go (hunk 6 above)
Comment (cstockton) [reply internal/e2e/e2ehooks/e2ehooks_test.go:51]: I disagree here, can we have a call to work through this?

File: internal/e2e/e2eapi/e2eapi_test.go (hunk 7 above)
Comment (cstockton) [reply internal/e2e/e2eapi/e2eapi_test.go]: I've added detailed comments for this.

File: internal/e2e/e2eapi/e2eapi_test.go [hunk 10]
```diff
@@ -130,46 +117,71 @@ func TestDo(t *testing.T) {
...
+			// handling in the switch statement within do.
+			testName := fmt.Sprintf("Status=%v", http.StatusText(statusCode))
+			t.Run(testName, func(t *testing.T) {
+
+				// We assign a sentinel error to ensure propagation.
+				sentinel := errors.New("sentinel")
+
+				// This implementation of the http.RoundTripper is a way to
+				// cover the io.ReadAll(io.LimitReader(...)) lines in the switch
+				// statements inside do().
```
Comment (hf) [inline internal/e2e/e2eapi/e2eapi_test.go:138]: The comments are much more helpful, yes.

I'm gonna approve, but just for future -- we should try to avoid specifying the "implementation" of the things we test. That's super fragile and if you change the code these tests either fail and you spend a lot of time figuring out why, or they become flaky -- wasting more time.

Comment (cstockton) [reply internal/e2e/e2eapi/e2eapi_test.go:138]: I have a slightly different opinion around dependency injection and view the fact that changing code causes tests to fail as a benefit.

But I understand where you are coming from and will avoid this when I can.
